File: AzanAlarm-RN/src/services/SettingsService.ts
import { AppSettings, PrayerCalculationMethod, JuristicMethod, AppTheme } from '../types';
import { storageService } from './StorageService';
import { STORAGE_KEYS } from '../constants/storage';

export interface SettingsService {
    getSettings(): Promise<AppSettings>;
    updateSettings(settings: AppSettings): Promise<void>;
    resetToDefaults(): Promise<void>;
}

const DEFAULT_SETTINGS: AppSettings = {
    calculationMethod: PrayerCalculationMethod.MuslimWorldLeague,
    juristicMethod: JuristicMethod.Shafii,
    audioTheme: 'default',
    is24HourFormat: false,
    enableNotifications: true,
    enableVibration: true,
    theme: AppTheme.System,
    language: 'en',
};

class SettingsServiceImpl implements SettingsService {
    /**
     * Get app settings from storage or return defaults
     */
    async getSettings(): Promise<AppSettings> {
        try {
            const settings = storageService.get<AppSettings>(STORAGE_KEYS.SETTINGS);

            if (!settings) {
                // First time - save defaults
                await this.updateSettings(DEFAULT_SETTINGS);
                return DEFAULT_SETTINGS;
            }

            // Merge with defaults to handle new settings added in updates
            return {
                ...DEFAULT_SETTINGS,
                ...settings,
            };
        } catch (error) {
            console.error('Failed to get settings:', error);
            return DEFAULT_SETTINGS;
        }
    }

    /**
     * Update app settings
     */
    async updateSettings(settings: AppSettings): Promise<void> {
        try {
            // Validate settings
            this.validateSettings(settings);

            // Save to storage
            storageService.set(STORAGE_KEYS.SETTINGS, settings);
        } catch (error) {
            console.error('Failed to update settings:', error);
            throw error;
        }
    }

    /**
     * Reset settings to defaults
     */
    async resetToDefaults(): Promise<void> {
        try {
            storageService.set(STORAGE_KEYS.SETTINGS, DEFAULT_SETTINGS);
        } catch (error) {
            console.error('Failed to reset settings:', error);
            throw error;
        }
    }

    /**
     * Validate settings object
     */
    private validateSettings(settings: AppSettings): void {
        if (!settings) {
            throw new Error('Settings object is required');
        }

        // Validate calculation method
        if (!Object.values(PrayerCalculationMethod).includes(settings.calculationMethod)) {
            throw new Error('Invalid calculation method');
        }

        // Validate juristic method
        if (!Object.values(JuristicMethod).includes(settings.juristicMethod)) {
            throw new Error('Invalid juristic method');
        }

        // Validate theme
        if (!Object.values(AppTheme).includes(settings.theme)) {
            throw new Error('Invalid theme');
        }
    }
}

export const settingsService = new SettingsServiceImpl();

File: internal/services/settings_service.go
// Package services contains business logic for the AzanAlarm application.
package services

import (
	"AzanAlarm/internal/models"
)

// SettingsService handles application settings.
type SettingsService struct {
	storage  *StorageService
	settings models.AppSettings
}

// NewSettingsService creates a new SettingsService instance.
func NewSettingsService(storage *StorageService) *SettingsService {
	ss := &SettingsService{
		storage:  storage,
		settings: models.DefaultSettings(),
	}

	// Load existing settings
	var savedSettings models.AppSettings
	if err := storage.Load("settings", &savedSettings); err == nil {
		ss.settings = savedSettings
	}

	return ss
}

// GetSettings returns the current application settings.
func (ss *SettingsService) GetSettings() models.AppSettings {
	return ss.settings
}

// SaveSettings saves the application settings.
func (ss *SettingsService) SaveSettings(settings models.AppSettings) error {
	ss.settings = settings
	return ss.storage.Save("settings", settings)
}

// ResetToDefaults resets settings to default values.
func (ss *SettingsService) ResetToDefaults() error {
	ss.settings = models.DefaultSettings()
	return ss.storage.Save("settings", ss.settings)
}

// UpdateCalculationMethod updates only the calculation method.
func (ss *SettingsService) UpdateCalculationMethod(method models.PrayerCalculationMethod) error {
	ss.settings.CalculationMethod = method
	return ss.storage.Save("settings", ss.settings)
}

// UpdateJuristicMethod updates only the juristic method.
func (ss *SettingsService) UpdateJuristicMethod(method models.JuristicMethod) error {
	ss.settings.JuristicMethod = method
	return ss.storage.Save("settings", ss.settings)
}

// UpdateTheme updates the application theme.
func (ss *SettingsService) UpdateTheme(theme models.AppTheme) error {
	ss.settings.Theme = theme
	return ss.storage.Save("settings", ss.settings)
}

// ToggleNotifications toggles notification enabling.
func (ss *SettingsService) ToggleNotifications(enable bool) error {
	ss.settings.EnableNotifications = enable
	return ss.storage.Save("settings", ss.settings)
}

// Toggle24HourFormat toggles between 12h and 24h time format.
func (ss *SettingsService) Toggle24HourFormat(enable bool) error {
	ss.settings.Is24HourFormat = enable
	return ss.storage.Save("settings", ss.settings)
}
